corehq/apps/locations/bulk_management.py

```diff
@@ -0,0 +1,652 @@
+            .format(code)
+            for code in unlisted_codes
+        ]
+
+    @memoized
+    def _check_unlisted_location_ids(self):
+        # count locations not listed in excel but are present in the domain now
+        old = self.old_collection.locations_by_id
+        listed = [l.location_id for l in self.all_listed_locations]
+        unlisted = set(old.keys()) - set(listed)
+
+        return [
+            "Location '{name} (id {id})' is not listed in the excel. All locations should be listed"
+            .format(name=old[location_id].name, id=location_id)
+            for location_id in unlisted
+        ]
+
+    @memoized
+    def _check_unknown_location_ids(self):
+        # count location_ids listed in the excel that are not found in the domain
+        if not self.old_collection:
+            return []
+        old = self.old_collection.locations_by_id
+        listed = {l.location_id: l for l in self.all_listed_locations if l.location_id}
+        unknown = set(listed.keys()) - set(old.keys())
+
+        return [
+            "Location 'id: {id}' is not found in your domain. It's listed in the sheet {type} at row {index}"
+            .format(id=l_id, type=listed[l_id].location_type, index=listed[l_id].index)
+            for l_id in unknown
+        ]
+
+    @memoized
+    def _validate_types_tree(self):
+        type_pairs = [(lt.code, lt.parent_code) for lt in self.location_types]
+        try:
+            assert_no_cycles(type_pairs)
+        except BadParentError as e:
+            return [
+                "Location Type '{}' refers to a parent which doesn't exist".format(code)
+                for code in e.affected_nodes
+            ]
+        except CycleError as e:
+            return [
+                "Location Type '{}' has a parentage that loops".format(code)
+                for code in e.affected_nodes
+            ]
+
+        allowed_from_codes, allowed_to_codes = expansion_validators(type_pairs)
+        errors = []
+        for lt in self.location_types:
+            if lt.expand_from and lt.expand_from not in allowed_from_codes(lt.code):
+                errors.append(
+                    "'{}' can't have '{}' as 'Expand From', valid options are '{}'"
+                    .format(lt.code, lt.expand_from, allowed_from_codes)
+                )
+
+            if lt.sync_to and lt.sync_to not in allowed_to_codes(lt.code):
+                errors.append(
+                    "'{}' can't have '{}' as 'Sync To', valid options are '{}'"
+                    .format(lt.code, lt.sync_to, allowed_to_codes)
+                )
+
+        return errors
+
+    @memoized
+    def _validate_location_tree(self):
+        errors = []
+
+        def _validate_location(location):
+            loc_type = self.types_by_code.get(location.location_type)
+            if not loc_type:
+                return "Location '{}' in sheet points to a to be deleted location-type '{}'".format(
```

Comment:
Yeah, it's not super clear here itself, but the non existent type would have already been caught in `_validate_type_tree`, so here it must be a to be deleted location. I will change the wording since its not very clear.